File: transektcount/src/main/java/com/wmstein/transektcount/ListSpeciesActivity.java
package com.wmstein.transektcount;

import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ScrollView;

import com.wmstein.transektcount.database.Count;
import com.wmstein.transektcount.database.CountDataSource;
import com.wmstein.transektcount.database.Head;
import com.wmstein.transektcount.database.HeadDataSource;
import com.wmstein.transektcount.database.Meta;
import com.wmstein.transektcount.database.MetaDataSource;
import com.wmstein.transektcount.database.Section;
import com.wmstein.transektcount.database.SectionDataSource;
import com.wmstein.transektcount.widgets.ListHeadWidget;
import com.wmstein.transektcount.widgets.ListMetaWidget;
import com.wmstein.transektcount.widgets.ListSpeciesWidget;
import com.wmstein.transektcount.widgets.ListSumWidget;

import java.util.List;

/****************************************************
 * ListSpeciesActivity shows list of counting results
 * Created by wmstein on 2016-03-15,
 * last edited on 2019-02-12
 */
public class ListSpeciesActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener
{
    private static String TAG = "transektcountListSpecAct"; // for future use
    TransektCountApplication transektCount;
    SharedPreferences prefs;

    LinearLayout spec_area;

    Head head;
    Meta meta;

    // preferences
    private boolean awakePref;
    private String sortPref;
    private boolean screenOrientL; // option for screen orientation

    // the actual data
    private CountDataSource countDataSource;
    private SectionDataSource sectionDataSource;
    private HeadDataSource headDataSource;
    private MetaDataSource metaDataSource;

    ListHeadWidget lhw;
    ListMetaWidget lmw;
    ListSumWidget lsw;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list_species);

        countDataSource = new CountDataSource(this);
        sectionDataSource = new SectionDataSource(this);
        headDataSource = new HeadDataSource(this);
        metaDataSource = new MetaDataSource(this);

        transektCount = (TransektCountApplication) getApplication();
        prefs = TransektCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        awakePref = prefs.getBoolean("pref_awake", true);
        sortPref = prefs.getString("pref_sort_sp", "none"); // sorted species list
        screenOrientL = prefs.getBoolean("screen_Orientation", false);

        if (screenOrientL)
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        }
        else
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        ScrollView listSpec_screen = findViewById(R.id.listSpecScreen);
        listSpec_screen.setBackground(transektCount.getBackground());

        //noinspection ConstantConditions
        getSupportActionBar().setTitle(getString(R.string.viewSpecTitle));

        spec_area = findViewById(R.id.listSpecLayout);

        if (awakePref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        if (awakePref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }

        // build Show Results screen
        spec_area.removeAllViews();
        loadData();
    }

    // fill ListSpeciesWidget with relevant counts and sections data
    public void loadData()
    {
        int summf = 0, summ = 0, sumf = 0, sump = 0, suml = 0, sumo = 0;
        int summfe = 0, summe = 0, sumfe = 0, sumpe = 0, sumle = 0, sumoe = 0;
        int sumInt, sumExt;

        headDataSource.open();
        metaDataSource.open();

        //load head and meta data
        head = headDataSource.getHead();
        meta = metaDataSource.getMeta();

        // display the editable transect No.
        lhw = new ListHeadWidget(this, null);
        lhw.setWidgetLNo(getString(R.string.transectnumber));
        lhw.setWidgetLNo1(head.transect_no);
        lhw.setWidgetLName(getString(R.string.inspector));
        lhw.setWidgetLName1(head.inspector_name);
        spec_area.addView(lhw);

        // display the editable meta data
        lmw = new ListMetaWidget(this, null);
        lmw.setWidgetLMeta1(getString(R.string.temperature));
        lmw.setWidgetLItem1(meta.tempe);
        lmw.setWidgetLMeta2(getString(R.string.wind));
        lmw.setWidgetLItem2(meta.wind);
        lmw.setWidgetLMeta3(getString(R.string.clouds));
        lmw.setWidgetLItem3(meta.clouds);
        lmw.setWidgetLDate1(getString(R.string.date));
        lmw.setWidgetLDate2(meta.date);
        lmw.setWidgetLTime1(getString(R.string.starttm));
        lmw.setWidgetLItem4(meta.start_tm);
        lmw.setWidgetLTime2(getString(R.string.endtm));
        lmw.setWidgetLItem5(meta.end_tm);
        spec_area.addView(lmw);

        // display all the sorted counts by adding them to listSpecies layout
        List<Count> specs; // List of species

        int sect_id;
        Section section;
        countDataSource.open();
        sectionDataSource.open();

        switch (sortPref)
        {
        case "names_alpha":
            specs = countDataSource.getAllCountsForSrtName();
            break;
        case "codes":
            specs = countDataSource.getAllCountsForSrtCode();
            break;
        default:
            specs = countDataSource.getAllCounts();
            break;
        }

        int spec_countf1i;
        int spec_countf2i;
        int spec_countf3i;
        int spec_countpi;
        int spec_countli;
        int spec_countei;
        int spec_countf1e;
        int spec_countf2e;
        int spec_countf3e;
        int spec_countpe;
        int spec_countle;
        int spec_countee;

        for (Count spec : specs)
        {
            ListSpeciesWidget widget = new ListSpeciesWidget(this, null);
            sect_id = widget.getSpec_sectionid(spec);
            section = sectionDataSource.getSection(sect_id);
            widget.setCount(spec, section);

            spec_countf1i = widget.getSpec_countf1i(spec);
            spec_countf2i = widget.getSpec_countf2i(spec);
            spec_countf3i = widget.getSpec_countf3i(spec);
            spec_countpi = widget.getSpec_countpi(spec);
            spec_countli = widget.getSpec_countli(spec);
            spec_countei = widget.getSpec_countei(spec);
            spec_countf1e = widget.getSpec_countf1e(spec);
            spec_countf2e = widget.getSpec_countf2e(spec);
            spec_countf3e = widget.getSpec_countf3e(spec);
            spec_countpe = widget.getSpec_countpe(spec);
            spec_countle = widget.getSpec_countle(spec);
            spec_countee = widget.getSpec_countee(spec);

            summf = summf + spec_countf1i;
            summ = summ + spec_countf2i;
            sumf = sumf + spec_countf3i;
            sump = sump + spec_countpi;
            suml = suml + spec_countli;
            sumo = sumo + spec_countei;
            summfe = summfe + spec_countf1e;
            summe = summe + spec_countf2e;
            sumfe = sumfe + spec_countf3e;
            sumpe = sumpe + spec_countpe;
            sumle = sumle + spec_countle;
            sumoe = sumoe + spec_countee;
        }

        sumInt = summf + summ + sumf + sump + suml + sumo;
        sumExt = summfe + summe + sumfe + sumpe + sumle + sumoe;

        // display the totals
        lsw = new ListSumWidget(this, null);
        lsw.setSum(summf, summ, sumf, sump, suml, sumo, summfe, summe, sumfe, sumpe, sumle, sumoe, sumInt, sumExt);
        spec_area.addView(lsw);

        for (Count spec : specs)
        {
            ListSpeciesWidget widget = new ListSpeciesWidget(this, null);
            sect_id = widget.getSpec_sectionid(spec);
            section = sectionDataSource.getSection(sect_id);
            widget.setCount(spec, section);
            spec_area.addView(widget);
        }
    }

    @Override
    protected void onPause()
    {
        super.onPause();

        // close the data sources
        headDataSource.close();
        metaDataSource.close();
        countDataSource.close();
        sectionDataSource.close();

        if (awakePref)
        {
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    @Override
    protected void onStop()
    {
        super.onStop();

        // close the data sources
        headDataSource.close();
        metaDataSource.close();
        countDataSource.close();
        sectionDataSource.close();

        if (awakePref)
        {
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    /***************/
    public void saveAndExit(View view)
    {
        super.finish();
    }

    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        ScrollView listSpec_screen = findViewById(R.id.listSpecScreen);
        listSpec_screen.setBackground(null);
        listSpec_screen.setBackground(transektCount.setBackground());
        awakePref = prefs.getBoolean("pref_awake", true);
        sortPref = prefs.getString("pref_sort_sp", "none"); // sorted species list
        screenOrientL = prefs.getBoolean("screen_Orientation", false);
    }

}

File: transektcount/src/main/java/com/wmstein/transektcount/ListSpeciesActivity.kt
package com.wmstein.transektcount

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.WindowManager
import android.widget.LinearLayout
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import com.wmstein.transektcount.database.Count
import com.wmstein.transektcount.database.CountDataSource
import com.wmstein.transektcount.database.Head
import com.wmstein.transektcount.database.HeadDataSource
import com.wmstein.transektcount.database.Meta
import com.wmstein.transektcount.database.MetaDataSource
import com.wmstein.transektcount.database.Section
import com.wmstein.transektcount.database.SectionDataSource
import com.wmstein.transektcount.widgets.ListHeadWidget
import com.wmstein.transektcount.widgets.ListMetaWidget
import com.wmstein.transektcount.widgets.ListSpeciesWidget
import com.wmstein.transektcount.widgets.ListSumWidget

/************************************************************
 * ListSpeciesActivity shows results list of counted Species,
 * uses ListSpeciesWidget, ListHeadWidget, ListMetaWidget,
 * ListSumWidget
 * Created by wmstein on 2016-03-15,
 * last edited in Java on 2022-04-30,
 * converted to Kotlin on 2023-07-17,
 * last edited on 2024-12-17
 */
class ListSpeciesActivity : AppCompatActivity() {
    private var transektCount: TransektCountApplication? = null

    private var specArea: LinearLayout? = null

    // Data
    private var countDataSource: CountDataSource? = null
    private var sectionDataSource: SectionDataSource? = null
    private var headDataSource: HeadDataSource? = null
    private var metaDataSource: MetaDataSource? = null
    private var head: Head? = null
    private var meta: Meta? = null
    private var lhw: ListHeadWidget? = null
    private var lmw: ListMetaWidget? = null
    private var lsw: ListSumWidget? = null

    // Preferences
    private var prefs = TransektCountApplication.getPrefs()
    private var awakePref = false
    private var outPref: String? = null
    private var brightPref = false

    @SuppressLint("SourceLockedOrientationActivity")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (MyDebug.dLOG) Log.i(TAG, "59, onCreate")

        transektCount = application as TransektCountApplication
        awakePref = prefs.getBoolean("pref_awake", true)
        outPref = prefs.getString("pref_csv_out", "species") // sort mode output
        brightPref = prefs.getBoolean("pref_bright", true)

        setContentView(R.layout.activity_list_species)

        // Set full brightness of screen
        if (brightPref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            val params = window.attributes
            params.screenBrightness = 1.0f
            window.attributes = params
        }

        headDataSource = HeadDataSource(this)
        sectionDataSource = SectionDataSource(this)
        metaDataSource = MetaDataSource(this)
        countDataSource = CountDataSource(this)

        supportActionBar!!.title = getString(R.string.viewSpecTitle)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        specArea = findViewById(R.id.listSpecLayout)

        if (awakePref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        // new onBackPressed logic
        onBackPressedDispatcher.addCallback(object :
            OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (MyDebug.dLOG) Log.i(TAG, "94, handleOnBackPressed")

                NavUtils.navigateUpFromSameTask(this@ListSpeciesActivity)
            }
        })
    }
    // End of onCreate()

    override fun onResume() {
        super.onResume()

        if (MyDebug.dLOG) Log.i(TAG, "105, onResume")

        if (awakePref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        headDataSource!!.open()
        metaDataSource!!.open()
        countDataSource!!.open()
        sectionDataSource!!.open()

        // build Show Results screen
        specArea!!.removeAllViews()
        loadData()
    }
    // End of onResume()

    // fill ListSpeciesWidget with relevant counts and sections data
    private fun loadData() {
        var summf = 0
        var summ = 0
        var sumf = 0
        var sump = 0
        var suml = 0
        var sumo = 0
        var summfe = 0
        var summe = 0
        var sumfe = 0
        var sumpe = 0
        var sumle = 0
        var sumoe = 0

        //load head and meta data
        head = headDataSource!!.head
        meta = metaDataSource!!.meta

        // display the editable transect No.
        lhw = ListHeadWidget(this, null)
        lhw!!.setWidgetLNo(getString(R.string.transectnumber))
        lhw!!.setWidgetLNo1(head!!.transect_no)
        lhw!!.setWidgetLName(getString(R.string.inspector))
        lhw!!.setWidgetLName1(head!!.inspector_name)
        specArea!!.addView(lhw)

        // display the editable meta data
        lmw = ListMetaWidget(this, null)
        lmw!!.setWidgetLTemp(getString(R.string.temperature))
        lmw!!.setWidgetLTemps(meta!!.temps)
        lmw!!.setWidgetLTempe(meta!!.tempe)
        lmw!!.setWidgetLWind(getString(R.string.wind))
        lmw!!.setWidgetLWinds(meta!!.winds)
        lmw!!.setWidgetLWinde(meta!!.winde)
        lmw!!.setWidgetLCloud(getString(R.string.clouds))
        lmw!!.setWidgetLClouds(meta!!.clouds)
        lmw!!.setWidgetLCloude(meta!!.cloude)
        lmw!!.setWidgetLDate1(getString(R.string.date))
        lmw!!.setWidgetLDate2(meta!!.date)
        lmw!!.setWidgetLTime1(getString(R.string.starttm))
        lmw!!.setWidgetLItem4(meta!!.start_tm)
        lmw!!.setWidgetLTime2(getString(R.string.endtm))
        lmw!!.setWidgetLItem5(meta!!.end_tm)
        lmw!!.setWidgetLNote1(getString(R.string.note))
        lmw!!.setWidgetLNote2(meta!!.note)
        specArea!!.addView(lmw)

        // display all the sorted counts by adding them to listSpecies layout
        var sect_id: Int
        var section: Section
        val specs: List<Count> // List of sorted species

        if (outPref.equals("sections")) {
            // sort criteria are section and name
            specs = countDataSource!!.allCountsForSrtSectionName
        } else {
            // sort criteria are name and section
            specs = countDataSource!!.allCountsForSrtNameSection
        }

        val sumSpec: Int = countDataSource!!.diffSpec // get number of different species
        var specCntf1i: Int
        var specCntf2i: Int
        var specCntf3i: Int
        var specCntpi: Int
        var specCntli: Int
        var specCntei: Int
        var specCntf1e: Int
        var specCntf2e: Int
        var specCntf3e: Int
        var specCntpe: Int
        var specCntle: Int
        var specCntee: Int

        for (spec in specs) {
            val widget = ListSpeciesWidget(this, null)
            sect_id = widget.getSpecSectionid(spec)
            section = sectionDataSource!!.getSection(sect_id)
            widget.setCount(spec, section)
            specCntf1i = widget.getSpecCountf1i(spec)
            specCntf2i = widget.getSpecCountf2i(spec)
            specCntf3i = widget.getSpecCountf3i(spec)
            specCntpi = widget.getSpecCountpi(spec)
            specCntli = widget.getSpecCountli(spec)
            specCntei = widget.getSpecCountei(spec)
            specCntf1e = widget.getSpecCountf1e(spec)
            specCntf2e = widget.getSpecCountf2e(spec)
            specCntf3e = widget.getSpecCountf3e(spec)
            specCntpe = widget.getSpecCountpe(spec)
            specCntle = widget.getSpecCountle(spec)
            specCntee = widget.getSpecCountee(spec)

            summf += specCntf1i
            summ += specCntf2i
            sumf += specCntf3i
            sump += specCntpi
            suml += specCntli
            sumo += specCntei
            summfe += specCntf1e
            summe += specCntf2e
            sumfe += specCntf3e
            sumpe += specCntpe
            sumle += specCntle
            sumoe += specCntee
        }

        val sumInt: Int = summf + summ + sumf + sump + suml + sumo // sum of internal counts
        val sumExt: Int = summfe + summe + sumfe + sumpe + sumle + sumoe // sum of external counts

        // display the totals
        lsw = ListSumWidget(this, null)
        lsw!!.setSum(
            summf,
            summ,
            sumf,
            sump,
            suml,
            sumo,
            summfe,
            summe,
            sumfe,
            sumpe,
            sumle,
            sumoe,
            sumInt,
            sumExt,
            sumSpec
        )
        specArea!!.addView(lsw)

        // display all counted soecies per section
        for (spec in specs) {
            val widget = ListSpeciesWidget(this, null)
            sect_id = widget.getSpecSectionid(spec)
            section = sectionDataSource!!.getSection(sect_id)
            widget.setCount(spec, section)
            specArea!!.addView(widget)
        }
    }
    // End of loadData()

    override fun onPause() {
        super.onPause()

        if (MyDebug.dLOG) Log.i(TAG, "267, onPause")

        // close the data sources
        headDataSource!!.close()
        metaDataSource!!.close()
        countDataSource!!.close()
        sectionDataSource!!.close()

        if (awakePref) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    companion object {
        private const val TAG = "ListSpecAct"
    }

}
